File: AdministrationServer/repository/src/main/java/by/delta/repository/impl/MessageRepositoryImpl.java
package by.delta.repository.impl;

import by.delta.model.Message;
import by.delta.specification.abstractspecification.AbstractCriteriaQuerySpecification;
import org.springframework.stereotype.Repository;

import javax.persistence.criteria.CriteriaBuilder;
import javax.persistence.criteria.CriteriaQuery;
import javax.persistence.criteria.Root;
import java.util.List;

@Repository
public class MessageRepositoryImpl extends AbstractRepository<Message> {

    @Override
    Message getById(long id) {
        return entityManager.find(Message.class, id);
    }

    @Override
    public List<Message> query(AbstractCriteriaQuerySpecification<Message> specification, int limit, int offset) {
        CriteriaBuilder cb = entityManager.getCriteriaBuilder();
        CriteriaQuery<Message> messageCriteriaQuery = cb.createQuery(Message.class);
        Root<Message> root = messageCriteriaQuery.from(Message.class);
        return entityManager.createQuery(specification.toQuery(messageCriteriaQuery, root, cb)).setMaxResults(limit).setFirstResult(offset).getResultList();
    }
}

File: AdministrationServer/repository/src/main/java/by/delta/repository/impl/MessageRepositoryImpl.kt
package by.delta.repository.impl

import by.delta.model.Message
import by.delta.specification.abstractspecification.AbstractCriteriaQuerySpecification
import by.delta.specification.abstractspecification.ICountSpecification
import org.springframework.stereotype.Repository

@Repository
open class MessageRepositoryImpl : AbstractRepository<Message>() {

    internal override fun getById(id: Long): Message {
        return entityManager.find(Message::class.java, id)
    }

    override fun query(specification: AbstractCriteriaQuerySpecification<Message>, limit: Int, offset: Int): List<Message> {
        val cb = entityManager.criteriaBuilder
        val messageCriteriaQuery = cb.createQuery(Message::class.java)
        val root = messageCriteriaQuery.from(Message::class.java)
        return entityManager.createQuery(specification.toQuery(messageCriteriaQuery, root, cb)).setMaxResults(limit).setFirstResult(offset).resultList
    }

    override fun countQuery(specification: ICountSpecification): Long {
        val cb = entityManager.criteriaBuilder
        val messageCriteriaQuery = cb.createQuery()
        val root = messageCriteriaQuery.from(Message::class.java)
        val criteriaQuery = specification.toQuery(messageCriteriaQuery, root, cb)
        return  entityManager.createQuery(criteriaQuery).singleResult.toString().toLong()
    }
}
